### ISSUE TITLE
Lincoln Index - Three parameter model

### ISSUE THREAD

**Issue by zjachc** [AllenDowney/ThinkBayes2#62]
I am receiving the following error when I attempt to compute the likelihood in the three parameter model:
for N, p0, p1 in joint3_pmf.index:

ValueError: not enough values to unpack (expected 3, got 2)

I think the issue is that when I attempt to use make_joint using p1 and the joint2_pmf, I do not get the triplet, but rather only the pair, mean p1 is not attaching and thus the multi-index only has 2 items instead of three.

Do you have any suggestions about how best to fix?

Thank you

**Comment by AllenDowney** [comment]
Hi. To make sure I understand the question, are you reporting an error in my code or asking for help debugging?

**Comment by zjachc** [comment]
Hi - I used to code as it was written in the book and received an error; however, I'm not sure if there is a code error or a different issue and wanted to see if you had a suggestion as to why the code won't work as written.

**Comment by zjachc** [comment]
Shapes of the arrays match what was expected in the book. As do the heads up until joint3_pmf. At that point, the triplet doesn't show up, only a pair. It seems there is some issue with make_joint, Pmf, and stack, which has worked fine until now.

**Comment by AllenDowney** [comment]
When you run the code in the notebook I linked, does it work?

**Comment by zjachc** [comment]
Yes, it does. I understand the problem is on my end.

Do you have any thoughts as to why this issue emerged in this instance of the three-parameter, but never before?

**Comment by AllenDowney** [comment]
If you display `joint3_pmf.index`, does it have three columns?

**Comment by zjachc** [comment]
Yes - here's the first row: 

MultiIndex([((0.0, 0.0),  32)

**Comment by zjachc** [comment]
MultiIndex([((0.0, 0.0),  32),
            ((0.0, 0.0),  37),
            ((0.0, 0.0),  42),
            ((0.0, 0.0),  47),
            ((0.0, 0.0),  52),
            ((0.0, 0.0),  57),
            ((0.0, 0.0),  62),
            ((0.0, 0.0),  67),
            ((0.0, 0.0),  72),
            ((0.0, 0.0),  77),
            ...
            ((1.0, 1.0), 302),
            ((1.0, 1.0), 307),
            ((1.0, 1.0), 312),
            ((1.0, 1.0), 317),
            ((1.0, 1.0), 322),
            ((1.0, 1.0), 327),
            ((1.0, 1.0), 332),
            ((1.0, 1.0), 337),
            ((1.0, 1.0), 342),
            ((1.0, 1.0), 347)],
           length=166464)

**Comment by AllenDowney** [comment]
Yup, that's a sequence of pairs, where the first of each pair is a pair, so that's not the same as a sequence of triples.

So either there's a difference between your code and mine or a difference between your environment and mine. I think it's unlikely to be an environment issue, so I'd look for a difference in the code.

**Comment by AllenDowney** [comment]
Maybe put it in a Colab notebook. That would eliminate the environmental explanation. In the process, you might find the problem. If not, you could share the notebook with me.

**Comment by zjachc** [comment]
I will do that.

**Comment by zjachc** [comment]
Here's the script:
# Three parameter model

# Priors
# params N, p0, p1, reuse prior_N

# prior_N
qs = np.arange(32, 350, step=5)
prior_N = make_uniform(qs, name='N')
prior_N.head(3)

# prior_p0, prior_p1
qs = np.linspace(0, 1, num=51)
prior_p0 = make_uniform(qs, name='p0')
prior_p1 = make_uniform(qs, name='p1')

prior_p0.head(3)
prior_p1.head(3)

# Assemble into joint prior w/ 3-dim
# put first two into DataFrame
joint2 = make_joint(prior_p0, prior_N)
joint2.shape
joint2.head(10)

# Stack and put result into a Pmf
joint2_pmf = Pmf(joint2.stack())
joint2_pmf.head(10)

# Use make_joint again to add in 3rd param
joint3 = make_joint(prior_p1, joint2_pmf)
joint3.shape

joint3.head(10)

# result is a DataFrame w/ values of N & p0 in a multi-index that goes down the rows
# and values of p1 in an index that goes across the columns

# apply stack again
joint3_pmf = Pmf(joint3.stack())
joint3_pmf.head(10)
joint3_pmf.index

joint3_pmf.shape

# Loop that computes the likelihoods
likelihood = joint3_pmf.copy()
observed = data.sum()
x = data.copy()

from scipy.stats import multinomial

for N, p0, p1 in joint3_pmf.index:
    x[0] = N - observed
    y = compute_probs(p0, p1)
    likelihood[N, p0, p1] = multinomial.pmf(x, N, y)

**Comment by zjachc** [comment]
It's an environment issue, works in Jupyter but not Spyder unfortunately.

**Comment by zjachc** [comment]
What's interesting is that this does not work in Jupyter but does work in Spyder.

The following was copied directly - no modification.

Result is that the contour plot won't render. However, Spyder did not have this issue.

observed = k01 + k10 + k11

for N, p in joint_pmf.index:
    k00 = N - observed
    x = [k00, k01, k10, k11]
    q = 1-p
    y = [q*q, q*p, p*q, p*p]
    likelihood[N, p] = multinomial.pmf(x, N, y)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [24], in <cell line: 3>()
      6 q = 1-p
      7 y = [q*q, q*p, p*q, p*p]
----> 8 likelihood[N, p] = multinomial.pmf(x, N, y)

TypeError: 'numpy.float64' object does not support item assignment
